### PR TITLE
Improve job handling through occ

### PR DESCRIPTION
**Pull request by come-nc**
- [x] Clean up code
- [x] Add function to list/search jobs
- [x] Fix background-job:execute for queued jobs

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

### CODE REVIEW

File: lib/private/BackgroundJob/JobList.php [hunk 1]
```diff
@@ -180,10 +166,26 @@ public function has($job, $argument) {
 	 * memory problems when creating too many instances.
 	 */
 	public function getAll() {
+		return $this->getJobs(null, null, 0);
+	}
+
+	public function getJobs($job, ?int $limit, int $offset): array {
```
Comment (nickvergessen) [inline lib/private/BackgroundJob/JobList.php:172]: You can now adjust the deprecation method on `getAll` to hint to this method?

File: lib/private/BackgroundJob/JobList.php [hunk 2]
```diff
@@ -292,15 +290,14 @@ public function getDetailsById(int $id): ?array {
...
+	 * @param array{class:class-string<IJob>, id:mixed, last_run:mixed, argument:string} $row
 	 */
-	private function buildJob($row) {
+	private function buildJob(array $row): ?IJob {
 		try {
 			try {
 				// Try to load the job as a service
 				/** @var IJob $job */
-				$job = \OC::$server->query($row['class']);
+				$job = \OC::$server->get($row['class']);
```
Comment (nickvergessen) [inline lib/private/BackgroundJob/JobList.php:300]:
```suggestion
				$job = \OCP\Server::get($row['class']);
```

File: lib/public/BackgroundJob/IJobList.php [hunk 3]
```diff
@@ -82,6 +82,15 @@ public function has($job, $argument);
 	 */
 	public function getAll();
 
+	/**
+	 * Get jobs matching the search
+	 *
+	 * @param \OCP\BackgroundJob\IJob|class-string<IJob>|null $job
+	 * @return \OCP\BackgroundJob\IJob[]
```
Comment (nickvergessen) [inline lib/public/BackgroundJob/IJobList.php:89]:
```suggestion
	 * @param IJob|class-string<IJob>|null $job
	 * @return IJob[]
```

File: lib/public/BackgroundJob/IJobList.php [hunk 4]
```diff
@@ -99,49 +108,42 @@ public function getNext(bool $onlyTimeSensitive = false): ?IJob;
...
 	 * @since 7.0.0
 	 */
 	public function setLastRun(IJob $job);
 
 	/**
 	 * set the run duration of $job
 	 *
-	 * @param IJob $job
-	 * @param $timeTaken
+	 * @param int $timeTaken
```
[the remark this answers is not in the document]
Comment (come-nc) [reply lib/public/BackgroundJob/IJobList.php:139]: Can I consider that IJobList is only implemented by us and add strict typing all around? That would be cleaner for sure but I was not sure it was possible.

**Review by CarlSchwan — APPROVED** [review]

**Review by ArtificialOwl — APPROVED** [review]

File: lib/public/BackgroundJob/IJobList.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (blizzz) [reply lib/public/BackgroundJob/IJobList.php:139]: so adding the hint or not?

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**